master/pkg/schemas/merge.go

```diff
@@ -0,0 +1,214 @@
+		if src.IsZero() {
+			return cpy(obj)
+		}
+	}
+
+	// Handle nil pointers by simply copying the src.
+	if obj.Kind() == reflect.Ptr && obj.IsZero() {
+		return cpy(src)
+	}
+
+	// If the object is Mergable, we only call Merge on it and return it as-is.
+	if mergeable, ok := obj.Addr().Interface().(Mergable); ok {
+		mergeable.Merge(src.Interface())
+		return obj
+	}
+
+	switch obj.Kind() {
+	case reflect.Ptr:
+		// We already checked for nil pointers, so just recurse on the Elem of the value.
+		obj.Elem().Set(merge(obj.Elem(), src.Elem(), name))
+
+	case reflect.Struct:
+		// Detect what to do with union fields.  There are 4 important cases:
+		//  1. src has a union member, obj does not -> recurse into that field.
+		//  2. src has a union member, obj has the same one -> recurse into that field.
+		//  3. src has a union member, obj has the different one -> do not recurse.
+		//  4. src has no union member -> recursing is a noop and doesn't matter
+		// Logically, this reduces to:
+		//   - if obj has a union member, src does not have the same one -> don't recurse.
+		//   - else -> recurse
+		recurseIntoUnion := true
+		for i := 0; i < src.NumField(); i++ {
+			structField := src.Type().Field(i)
+			if _, ok := structField.Tag.Lookup("union"); ok {
+				if !obj.Field(i).IsZero() && src.Field(i).IsZero() {
+					recurseIntoUnion = false
+					break
+				}
+			}
+		}
+		// Recurse into each field of the struct.
+		for i := 0; i < src.NumField(); i++ {
+			structField := src.Type().Field(i)
+			if _, ok := structField.Tag.Lookup("union"); ok && !recurseIntoUnion {
+				continue
+			}
+			fieldName := fmt.Sprintf("%v.%v", name, structField.Name)
+			x := merge(obj.Field(i), src.Field(i), fieldName)
+			obj.Field(i).Set(x)
+		}
+
+	case reflect.Map:
+		// Maps get fused together; all input keys are written into the output map.
+		for _, key := range src.MapKeys() {
+			// Ensure key is not already set in obj.
+			if objVal := obj.MapIndex(key); objVal.IsValid() {
+				continue
+			}
+			val := src.MapIndex(key)
+			obj.SetMapIndex(key, cpy(val))
+		}
+
+	case reflect.Slice:
+		// Slices get copied only if the original was a nil pointer, which should always pass
+		// through the cpy() codepath and never through here.
+
+	// Assert that none of the "complex" kinds are present.
+	case reflect.Array,
+		reflect.Chan,
+		reflect.Func,
+		reflect.Interface,
+		reflect.UnsafePointer:
+		panic(fmt.Sprintf("unable to fill %T with %T at %v", obj.Interface(), src.Interface(), name))
+
+		// Nothing to do for the simple Kinds like string or int; the only way a simple kind in the
+		// src can end up being merged into the obj is if it is within a call to cpy(), like after
+		// allocating a new pointer.  This is because we only merge into nil pointers.
+	}
+
+	return obj
+}
+
+// cpy is for deep copying, but it will only work on "nice" objects, which should include our
+// schema objects.
+func cpy(v reflect.Value) reflect.Value {
```

Comment:
I know you hate it but I'm a fan of the json unmarshal/marshal trick to do this.

Reply:
Yes, I still hate it.  Copying should involve writing values to strings and then parsing those strings.  json unmarshal/marshal might be more flexible for generic objects, but since we are only operating on a subset of all possible objects here, I'd prefer this.